### ISSUE TITLE
Possible CUDA memory fragmentation in transducer decoding

### ISSUE THREAD

**Issue by pzelasko** [k2-fsa/icefall#317]
I noticed that during decoding of `pruned_transducer_stateless` with `greedy_search_batch` method that CUDA memory seems to get fragmented with OOM exceptions like these:

```
RuntimeError: CUDA out of memory. Tried to allocate 7.28 GiB (GPU 0; 22.20 GiB total capacity; 7.80 GiB already allocated; 6.15 GiB free; 14.18 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```

Note that here, reserved memory (14GB) is >> than allocated memory (7.8GB). After setting the following `export PYTORCH_CUDA_ALLOC_CONF=max_split_size_mb:128` I was able to continue decoding without issues (so far). I can add that decoding is performed on out-of-domain data which is much noisier; I did not observe this behavior previously on clean data such as LibriSpeech.

Do you have any ideas where the fragmentation may be coming from?

**Comment by csukuangfj** [comment]
Are there large differences in duration across your test batches? The greedy search batch should not use too much GPU memory. The only part I can think of is the forward pass of the encoder model; if your decoding batches have very different duration distributions, it may cause memory fragmentation.

**Comment by pzelasko** [comment]
Yes, I expect the durations to be varying a lot. The crash is indeed from the forward pass of the encoder. Do you think there is anything that can be done about it?

**Comment by csukuangfj** [comment]
How about using a bucketing sampler so as to make batches have similar max durations?

**Comment by pzelasko** [comment]
Already using it; I thought you were referring to the variance of max sequence durations across batches.

**Comment by csukuangfj** [comment]
The output shape of the encoder is (N, T, C). If you can make different batches to have a similar `N*T`, then it may alleviate the fragmentation issue.

**Comment by pzelasko** [comment]
Hmm, I see. Maybe tweaking the bucketing settings can help a bit but I'm decoding quite various data, it'd require to find a setup-specific setting for each domain I guess. Anyway setting the env var helps to some extent at least. Thanks!

**Comment by csukuangfj** [comment]
From https://pytorch.org/docs/stable/notes/cuda.html#memory-management  , it may affect performance.
> Performance cost can range from ‘zero’ to ‘substatial’ depending on allocation patterns. Default value is unlimited, i.e. all blocks can be split. The [memory_stats()](https://pytorch.org/docs/stable/generated/torch.cuda.memory_stats.html#torch.cuda.memory_stats) and [memory_summary()](https://pytorch.org/docs/stable/generated/torch.cuda.memory_summary.html#torch.cuda.memory_summary) methods are useful for tuning. This option should be used as a last resort for a workload that is aborting due to ‘out of memory’ and showing a large amount of inactive split blocks.

**Comment by pzelasko** [comment]
I am tracking the RTF, it's around 700-1000x, I didn't notice it decreasing too much.

**Comment by csukuangfj** [comment]
ok. By the way, are the WERs of your datasets trained with pruned RNN-T comparable to the one trained with conformer_ctc, and also does it decode faster than conformer_ctc ?

**Comment by pzelasko** [comment]
I don't have a direct apple-to-apple comparison; I previously trained a pure-CTC conformer (without attention decoder in training). The WERs of pruned RNN-T are better than pure-CTC conformer by ~15% relative. The decoding time of greedy RNN-T with no LM seems to be in the range of 2-3x faster than using the pure-CTC conformer with 1best decoding over 4-gram LM.

**Comment by csukuangfj** [comment]
thanks
